feat: add conversion to v2 datamodel (DSP-1419)

BalduinLandolt opened the pull request:
resolves DSP-1419

This PR adds a conversion script to transform a metadata file in .ttl to a json file according to the new data model.  
It also adds a menu item to the GUI that prompts a dialog, which allows the user to run this conversion on one or many file(s).

irinaschubert commented on README.md:
```diff
@@ -58,6 +60,34 @@ All data is locally stored in the file `~/DaSCH/config/repos.data`. for more det
+This corresponds to the data that can currently be viewed in the [DaSCH Metadata Browser](https://meta.dasch.swiss).
+
+The initial datamodel will eventually be replaced by the model V2 which introduces major improvements.  
+Metadata V2 will evbentually be collected directly in the web interface rather than in this python application.  
```
typo "evbentually"

BalduinLandolt replied:
done, thanks.

irinaschubert commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+This corresponds to the data that can currently be viewed in the [DaSCH Metadata Browser](https://meta.dasch.swiss).
+
+The initial datamodel will eventually be replaced by the model V2 which introduces major improvements.  
+Metadata V2 will evbentually be collected directly in the web interface rather than in this python application.  
```
```suggestion
Metadata V2 will eventually be collected directly in the web interface rather than in this python application.  
```

irinaschubert commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+As output simply select a folder. The selected input file(s) will be converted and saved to this folder.  
+The output files will retain the input file's name but the ending will be replaced from `.ttl` to `.json`.
+If the output folder already contains a file whith such a name, it will get overwritten.
+It is therefor advisable to select an empty folder.
```
```suggestion
It is therefore advisable to select an empty folder.
```

BalduinLandolt replied:
done

irinaschubert commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+It is therefor advisable to select an empty folder.
+
+Finally, hit the `Convert` button at the very bottom of the dialog.  
+This button is greyed out als long as either the input or the output is missing. As soon as input and output are specified, the button gets enabled.
```
```suggestion
This button is greyed out as long as either the input or the output is missing. As soon as input and output are specified, the button gets enabled.
```

BalduinLandolt replied:
done

irinaschubert commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+Once it's done the output folder should open up, and the conversion dialog can be closed.  
+The process should not take more than a minute per file. On the console you should get minimal indication of the progress.
+
+**NB:** The conversion requires a local `.ttl` file. It is therefor necessary to export the data after entering it, before it can be converted.
```
```suggestion
**NB:** The conversion requires a local `.ttl` file. It is therefor enecessary to export the data after entering it, before it can be converted.
```

BalduinLandolt replied:
done

irinaschubert commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+
+For convenience, the conversion adds the string `XX` wherever the output can not be determined with sufficient confidence.
+__Please check those instances manually.__  
+Simply open the output file in a terxt editor and search for `XX`.
```
```suggestion
Simply open the output file in a text editor and search for `XX`.
```

BalduinLandolt replied:
done

irinaschubert commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+As the model V2 supports JSON-Schema validation, it is also advisable to open the output file in a text editor that supports JSON Schema validation.
+(E.g. Visual Studio Code which automatically validates JSON files that have a schema specified.)  
+This will highlight potential issues with the data.  
+> Note: The schema validates URLs as URIs. It will therefor consider URLs without scheme as invalid,
```
```suggestion
> Note: The schema validates URLs as URIs. It will therefore consider URLs without scheme as invalid,
```

BalduinLandolt replied:
done

irinaschubert commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+If the output can be determined with a sufficient level of confidence, the conversion will ___not___ add `XX`.
+__However it is still advisable to check the entirety of the output for potential errors.__
+
+As the model V2 supports JSON-Schema validation, it is also advisable to open the output file in a text editor that supports JSON Schema validation.
```
```suggestion
As the model V2 supports JSON schema validation, it is also advisable to open the output file in a text editor that supports JSON schema validation.
```

BalduinLandolt replied:
done

irinaschubert commented on dspMetadataGUI/util/converter.py:
```diff
@@ -0,0 +1,763 @@
+        data (str): string of a turtle setrialization of metadata
+
+    Returns:
+        str: json serialized metadata.
```
```suggestion
        str: json serialized metadata
```

irinaschubert commented on dspMetadataGUI/util/converter.py:
```diff
@@ -0,0 +1,763 @@
+
+    for _, p, o in g.triples((grant_iri, None, None)):
+        obj = str(o)
+        if p == dsp.hasFunder:
+            res.setdefault('funders', [])
+            res['funders'].append(obj)
+        elif p == dsp.hasNumber:
+            res['number'] = obj
+        elif p == dsp.hasName:
+            res['name'] = obj
+        elif p == dsp.hasURL:
+            res['url'] = _get_url(g, o)
+        # default cases
+        elif p == RDF.type:
+            pass
+        else:
+            print("Issue: Could not handle in Grant:", p, obj)
```
a switch-case structure could be used instead of if/elif/else

irinaschubert commented on dspMetadataGUI/util/converter.py:
```diff
@@ -0,0 +1,763 @@
+            res['keywords'].append(_guess_language_of_text(obj))
+        elif p == dsp.hasDiscipline:
+            res.setdefault('disciplines', [])
+            if isinstance(o, BNode):
+                res['disciplines'].append(_get_url(g, o))
+            else:
+                res['disciplines'].append(_guess_language_of_text(obj))
+        elif p == dsp.hasTemporalCoverage:
+            res.setdefault('temporalCoverage', [])
+            if isinstance(o, BNode):
+                res['temporalCoverage'].append(_get_url(g, o))
+            else:
+                res['temporalCoverage'].append(_guess_language_of_text(obj))
+        elif p == dsp.hasSpatialCoverage:
+            res.setdefault('spatialCoverage', [])
+            res['spatialCoverage'].append(_get_place(g, o))
+        elif p == dsp.hasURL:
+            res.setdefault('urls', [])
+            res['urls'].append(_get_url(g, o))
+        elif p == dsp.hasDataManagementPlan:
+            res['dataManagementPlan'] = obj
+        elif p == dsp.hasPublication:
+            res.setdefault('publications', [])
+            res['publications'].append(obj)
+        elif p == dsp.hasGrant:
+            res.setdefault('grants', [])
+            res['grants'].append(obj)
+        elif p == dsp.hasAlternateName:
+            res.setdefault('alternativeNames', [])
+            res['alternativeNames'].append(_guess_language_of_text(obj))
+        elif p == dsp.hasFunder:
+            res.setdefault('funders', [])
+            res['funders'].append(obj)
+        elif p == dsp.hasContactPoint:
+            res['contactPoint'] = obj
+        # default cases
+        elif p == RDF.type:
+            pass
+        else:
+            print("Issue: Could not handle in Project:", p, obj)
```
a switch-case structure could be used instead of if/elif/else

irinaschubert commented on dspMetadataGUI/util/converter.py:
```diff
@@ -0,0 +1,763 @@
+    #          ]
+    files = ['drawings-gods.ttl', 'awg.ttl']
+    results = {}
+    if not os.path.exists('out'):
+        os.mkdir('out')
```
Gemäss https://stackoverflow.com/questions/273192/how-can-i-safely-create-a-nested-directory könnte es hier zu einer Race Condition kommen. Alternativ könnte man folgenden Code benutzen:

`os.makedirs("path/to/directory", exist_ok=True)`

Oder:

```
from pathlib import Path
Path("/my/directory").mkdir(parents=True, exist_ok=True)
```

BalduinLandolt replied:
Interesting! I wasn't aware of that issue, thanks for pointing it out!

BalduinLandolt commented on README.md:
```diff
@@ -58,6 +60,34 @@ All data is locally stored in the file `~/DaSCH/config/repos.data`. for more det
+
+The most important changes from V1 to V2 include the following additions:
+
+- Support for multi-language literals.
```
done

BalduinLandolt commented on README.md:
```diff
@@ -58,6 +60,34 @@ All data is locally stored in the file `~/DaSCH/config/repos.data`. for more det
+
+- `howToCite` on project level
+
+- `country` property for addresses.
```
done

BalduinLandolt commented on README.md:
```diff
@@ -58,6 +60,34 @@ All data is locally stored in the file `~/DaSCH/config/repos.data`. for more det
+
+- `country` property for addresses.
+
+- Creation and modification timestamps.
```
done

BalduinLandolt commented on README.md:
```diff
@@ -58,6 +60,34 @@ All data is locally stored in the file `~/DaSCH/config/repos.data`. for more det
+
+- Creation and modification timestamps.
+
+- JSON-Schema validation
```
done

BalduinLandolt commented on docs/conversion.md:
```diff
@@ -0,0 +1,85 @@
+The metadata generated by the application conforms to the first version of the data model for metadata.  
+This corresponds to the data that can currently be viewed in the [DaSCH Metadata Browser](https://meta.dasch.swiss).
+
+The initial datamodel will eventually be replaced by the model V2 which introduces major improvements.  
```
done

BalduinLandolt commented on dspMetadataGUI/collectMetadata.py:
```diff
@@ -1328,5 +1338,141 @@ def on_key_down(self, evt):
+
+    def _on_select_input(self, label: wx.StaticText, mode):
+        """Select input file(s) by opening FileDialog/DirDialog, depending on the selected mode."""
+        if mode == ConverterDialog.InType.SINGLE_FILE:
+            with wx.FileDialog(self, "Select Input File", wildcard="*.ttl") as fileDialog:
+                res = fileDialog.ShowModal()
+                if res == wx.ID_OK:
+                    path = fileDialog.GetPath()
+                    self.__in_files = [path]
+                    label.SetLabel(str(self.__in_files))
+        elif mode == ConverterDialog.InType.MULTI_FILE:
+            with wx.FileDialog(self, "Select Input Files", style=wx.FD_MULTIPLE, wildcard="*.ttl") as fileDialog:
+                res = fileDialog.ShowModal()
+                if res == wx.ID_OK:
+                    paths = fileDialog.GetPaths()
+                    self.__in_files = paths
+                    if len(self.__in_files) < 6:
+                        label.SetLabel("\n".join(self.__in_files))
+                    else:
+                        label.SetLabel(f"Selected files: {len(self.__in_files)}")
+        elif mode == ConverterDialog.InType.DIRECTORY:
+            with wx.DirDialog(self, "Select Input File") as dirDialog:
+                res = dirDialog.ShowModal()
+                if res == wx.ID_OK:
+                    path = dirDialog.GetPath()
+                    files = glob(f'{path}/*.ttl')
+                    self.__in_files = files
+                    if len(self.__in_files) < 6:
+                        label.SetLabel("\n".join(self.__in_files))
+                    else:
+                        label.SetLabel(f"Selected files: {len(self.__in_files)}")
```
Probably this could be simplified, however I'd actually rather leave it like that because I found that simplifying things relating to `wxPython` is often a lot harder than it should be... and the present solution works at least.

BalduinLandolt commented:
> It looks good to me, I added some suggestions for minor code improvements and typos

thanks! implemented most changes, except the switch-case thing (as we discussed) and the simplification of the file dialogs (I don't trust wxPython enough ^^).

If you approve, I'll merge
